File: packages/tools/src/lib/msix.ts
import { cp, mkdir, readdir, readFile, rm, writeFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { deflateSync } from "node:zlib";
import { execa } from "execa";

export type MsixArchitecture = "x64" | "arm64";

export type MsixIdentityOptions = {
  displayName?: string;
  identityName: string;
  publisher: string;
  publisherDisplayName: string;
};

export type MsixBuildOptions = {
  arch: MsixArchitecture;
  executablePath: string;
  outputPath?: string;
  packageRoot?: string;
  repoRoot: string;
};

export type MsixBundleOptions = {
  outputPath?: string;
  packageDir?: string;
  repoRoot: string;
};

const APP_DISPLAY_NAME = "proxer";
const APP_DESCRIPTION =
  "Git-backed configuration synchronization tool for dotfiles";
const APP_EXECUTABLE_NAME = "proxer.exe";
const DEFAULT_LANGUAGE = "en-US";
const MIN_WINDOWS_VERSION = "10.0.19041.0";
const MAX_TESTED_WINDOWS_VERSION = "10.0.26100.0";

const resolveRepoPath = (repoRoot: string, path: string): string => {
  return resolve(repoRoot, path);
};

const readRequiredEnv = (env: NodeJS.ProcessEnv, name: string): string => {
  const value = env[name];

  if (value === undefined || value.trim() === "") {
    throw new Error(`${name} is required to build Windows MSIX packages`);
  }

  return value;
};

export const readMsixIdentityFromEnv = (
  env: NodeJS.ProcessEnv = process.env,
): MsixIdentityOptions => {
  // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
  const displayName = env["MSIX_DISPLAY_NAME"];

  return {
    identityName: readRequiredEnv(env, "MSIX_IDENTITY_NAME"),
    publisher: readRequiredEnv(env, "MSIX_PUBLISHER"),
    publisherDisplayName: readRequiredEnv(env, "MSIX_PUBLISHER_DISPLAY_NAME"),
    ...(displayName !== undefined && displayName.trim() !== ""
      ? { displayName }
      : {}),
  };
};

export const convertVersionToMsixVersion = (version: string): string => {
  const coreVersion = version.split("-")[0];

  if (coreVersion === undefined) {
    throw new Error(`Invalid package version for MSIX: ${version}`);
  }

  const parts = coreVersion.split(".");

  if (parts.length !== 3) {
    throw new Error(`Invalid package version for MSIX: ${version}`);
  }

  const numericParts = parts.map((part) => {
    if (!/^\d+$/.test(part)) {
      throw new Error(`Invalid package version for MSIX: ${version}`);
    }

    const value = Number(part);

    if (!Number.isSafeInteger(value) || value < 0 || value > 65_535) {
      throw new Error(`MSIX version segment out of range: ${part}`);
    }

    return value;
  });

  return `${numericParts[0]}.${numericParts[1]}.${numericParts[2]}.0`;
};

const readCliPackageVersion = async (repoRoot: string): Promise<string> => {
  const packageJsonPath = join(repoRoot, "packages", "cli", "package.json");
  const packageJson = JSON.parse(await readFile(packageJsonPath, "utf8")) as {
    version?: unknown;
  };

  if (typeof packageJson.version !== "string") {
    throw new Error(`Missing version in ${packageJsonPath}`);
  }

  return convertVersionToMsixVersion(packageJson.version);
};

const escapeXml = (value: string): string => {
  return value
    .replaceAll("&", "&amp;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&apos;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;");
};

export const buildMsixManifest = (options: {
  arch: MsixArchitecture;
  identity: MsixIdentityOptions;
  version: string;
}): string => {
  const displayName = options.identity.displayName ?? APP_DISPLAY_NAME;

  return `<?xml version="1.0" encoding="utf-8"?>
<Package
  xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
  xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
  xmlns:uap5="http://schemas.microsoft.com/appx/manifest/uap/windows10/5"
  xmlns:uap10="http://schemas.microsoft.com/appx/manifest/uap/windows10/10"
  xmlns:desktop4="http://schemas.microsoft.com/appx/manifest/desktop/windows10/4"
  xmlns:rescap="http://schemas.microsoft.com/appx/manifest/foundation/windows10/restrictedcapabilities"
  IgnorableNamespaces="uap uap5 uap10 desktop4 rescap">
  <Identity
    Name="${escapeXml(options.identity.identityName)}"
    Publisher="${escapeXml(options.identity.publisher)}"
    Version="${escapeXml(options.version)}"
    ProcessorArchitecture="${options.arch}" />
  <Properties>
    <DisplayName>${escapeXml(displayName)}</DisplayName>
    <PublisherDisplayName>${escapeXml(options.identity.publisherDisplayName)}</PublisherDisplayName>
    <Logo>Assets\\StoreLogo.png</Logo>
  </Properties>
  <Resources>
    <Resource Language="${DEFAULT_LANGUAGE}" />
  </Resources>
  <Dependencies>
    <TargetDeviceFamily
      Name="Windows.Desktop"
      MinVersion="${MIN_WINDOWS_VERSION}"
      MaxVersionTested="${MAX_TESTED_WINDOWS_VERSION}" />
  </Dependencies>
  <Capabilities>
    <rescap:Capability Name="runFullTrust" />
  </Capabilities>
  <Applications>
    <Application
      Id="proxer"
      Executable="${APP_EXECUTABLE_NAME}"
      EntryPoint="Windows.FullTrustApplication"
      uap10:RuntimeBehavior="packagedClassicApp"
      uap10:TrustLevel="mediumIL"
      desktop4:SupportsMultipleInstances="true">
      <uap:VisualElements
        DisplayName="${escapeXml(displayName)}"
        Description="${escapeXml(APP_DESCRIPTION)}"
        Square150x150Logo="Assets\\Square150x150Logo.png"
        Square44x44Logo="Assets\\Square44x44Logo.png"
        BackgroundColor="#102A43" />
      <Extensions>
        <uap5:Extension Category="windows.appExecutionAlias">
          <uap5:AppExecutionAlias desktop4:Subsystem="console">
            <uap5:ExecutionAlias Alias="${APP_EXECUTABLE_NAME}" />
          </uap5:AppExecutionAlias>
        </uap5:Extension>
      </Extensions>
    </Application>
  </Applications>
</Package>
`;
};

const crcTable = Array.from({ length: 256 }, (_, index) => {
  let crc = index;

  for (let bit = 0; bit < 8; bit++) {
    crc = (crc & 1) === 1 ? 0xedb8_8320 ^ (crc >>> 1) : crc >>> 1;
  }

  return crc >>> 0;
});

const crc32 = (buffers: Buffer[]): number => {
  let crc = 0xffff_ffff;

  for (const buffer of buffers) {
    for (const byte of buffer) {
      const tableValue = crcTable[(crc ^ byte) & 0xff];

      if (tableValue === undefined) {
        throw new Error("CRC table lookup failed");
      }

      crc = (crc >>> 8) ^ tableValue;
    }
  }

  return (crc ^ 0xffff_ffff) >>> 0;
};

const pngChunk = (type: string, data: Buffer): Buffer => {
  const typeBuffer = Buffer.from(type, "ascii");
  const length = Buffer.alloc(4);
  const crc = Buffer.alloc(4);

  length.writeUInt32BE(data.length, 0);
  crc.writeUInt32BE(crc32([typeBuffer, data]), 0);

  return Buffer.concat([length, typeBuffer, data, crc]);
};

const createLogoPng = (size: number): Buffer => {
  const bytesPerPixel = 4;
  const rowSize = 1 + size * bytesPerPixel;
  const raw = Buffer.alloc(rowSize * size);
  const base = [16, 42, 67, 255] as const;
  const accent = [20, 184, 166, 255] as const;
  const light = [245, 247, 250, 255] as const;
  const bandStart = Math.floor(size * 0.22);
  const bandEnd = Math.floor(size * 0.34);
  const markStart = Math.floor(size * 0.42);
  const markEnd = Math.floor(size * 0.58);

  for (let y = 0; y < size; y++) {
    const rowOffset = y * rowSize;
    raw[rowOffset] = 0;

    for (let x = 0; x < size; x++) {
      const pixelOffset = rowOffset + 1 + x * bytesPerPixel;
      const color =
        (x >= bandStart && x <= bandEnd) ||
        (x >= markStart && x <= markEnd && y >= markStart && y <= markEnd)
          ? accent
          : x >= markStart && x <= markEnd
            ? light
            : base;

      raw[pixelOffset] = color[0];
      raw[pixelOffset + 1] = color[1];
      raw[pixelOffset + 2] = color[2];
      raw[pixelOffset + 3] = color[3];
    }
  }

  const signature = Buffer.from([
    0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a,
  ]);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(size, 0);
  ihdr.writeUInt32BE(size, 4);
  ihdr[8] = 8;
  ihdr[9] = 6;
  ihdr[10] = 0;
  ihdr[11] = 0;
  ihdr[12] = 0;

  return Buffer.concat([
    signature,
    pngChunk("IHDR", ihdr),
    pngChunk("IDAT", deflateSync(raw)),
    pngChunk("IEND", Buffer.alloc(0)),
  ]);
};

const writeMsixAssets = async (assetsDirectory: string): Promise<void> => {
  await mkdir(assetsDirectory, { recursive: true });

  await Promise.all([
    writeFile(join(assetsDirectory, "StoreLogo.png"), createLogoPng(50)),
    writeFile(join(assetsDirectory, "Square44x44Logo.png"), createLogoPng(44)),
    writeFile(
      join(assetsDirectory, "Square150x150Logo.png"),
      createLogoPng(150),
    ),
  ]);
};

const findWindowsSdkTool = async (toolName: string): Promise<string> => {
  const envOverride = process.env[`${toolName.toUpperCase()}_PATH`];

  if (envOverride !== undefined && envOverride.trim() !== "") {
    return envOverride;
  }

  if (process.platform !== "win32") {
    throw new Error(`${toolName}.exe is only available on Windows runners`);
  }

  // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
  const windowsSdkDir = process.env["WindowsSdkDir"];
  // biome-ignore lint/complexity/useLiteralKeys: must use bracket notation for index signature access
  const programFiles = process.env["ProgramFiles"];
  const programFilesX86 = process.env["ProgramFiles(x86)"];

  const sdkRoots = [
    windowsSdkDir !== undefined ? join(windowsSdkDir, "bin") : undefined,
    programFilesX86 !== undefined
      ? join(programFilesX86, "Windows Kits", "10", "bin")
      : undefined,
    programFiles !== undefined
      ? join(programFiles, "Windows Kits", "10", "bin")
      : undefined,
  ].filter((value): value is string => value !== undefined);

  for (const sdkRoot of sdkRoots) {
    let versions: string[];

    try {
      versions = (await readdir(sdkRoot, { withFileTypes: true }))
        .filter((entry) => entry.isDirectory())
        .map((entry) => entry.name)
        .sort((left, right) =>
          right.localeCompare(left, undefined, { numeric: true }),
        );
    } catch {
      continue;
    }

    for (const version of versions) {
      for (const arch of ["x64", "x86"]) {
        const candidate = join(sdkRoot, version, arch, `${toolName}.exe`);

        try {
          await readFile(candidate);
          return candidate;
        } catch {}
      }
    }
  }

  return `${toolName}.exe`;
};

const createPriFile = async (packageRoot: string): Promise<void> => {
  const makePriPath = await findWindowsSdkTool("makepri");
  const priConfigPath = join(packageRoot, "priconfig.xml");

  await execa(
    makePriPath,
    ["createconfig", "/cf", priConfigPath, "/dq", DEFAULT_LANGUAGE],
    {
      cwd: packageRoot,
      stdio: "inherit",
    },
  );

  await execa(makePriPath, ["new", "/pr", packageRoot, "/cf", priConfigPath], {
    cwd: packageRoot,
    stdio: "inherit",
  });
  await rm(priConfigPath, { force: true });
};

export const performMsixBuild = async (
  options: MsixBuildOptions,
): Promise<{ outputPath: string; packageRoot: string }> => {
  const identity = readMsixIdentityFromEnv();
  const version = await readCliPackageVersion(options.repoRoot);
  const packageRoot =
    options.packageRoot !== undefined
      ? resolveRepoPath(options.repoRoot, options.packageRoot)
      : join(options.repoRoot, "packages", "cli", "dist", "msix", options.arch);
  const outputPath =
    options.outputPath !== undefined
      ? resolveRepoPath(options.repoRoot, options.outputPath)
      : join(
          options.repoRoot,
          "packages",
          "cli",
          "dist",
          "pkg",
          `proxer-win-${options.arch}.msix`,
        );
  const sourceExecutablePath = resolveRepoPath(
    options.repoRoot,
    options.executablePath,
  );

  await rm(packageRoot, { force: true, recursive: true });
  await mkdir(packageRoot, { recursive: true });
  await mkdir(dirname(outputPath), { recursive: true });
  await cp(sourceExecutablePath, join(packageRoot, APP_EXECUTABLE_NAME));
  await writeFile(
    join(packageRoot, "AppxManifest.xml"),
    buildMsixManifest({
      arch: options.arch,
      identity,
      version,
    }),
  );
  await writeMsixAssets(join(packageRoot, "Assets"));
  await createPriFile(packageRoot);

  const makeAppxPath = await findWindowsSdkTool("makeappx");
  await execa(
    makeAppxPath,
    ["pack", "/v", "/o", "/h", "SHA256", "/d", packageRoot, "/p", outputPath],
    {
      cwd: options.repoRoot,
      stdio: "inherit",
    },
  );

  console.log(`MSIX package generated at ${outputPath}`);

  return { outputPath, packageRoot };
};

export const performMsixBundle = async (
  options: MsixBundleOptions,
): Promise<{ outputPath: string }> => {
  const version = await readCliPackageVersion(options.repoRoot);
  const packageDir =
    options.packageDir !== undefined
      ? resolveRepoPath(options.repoRoot, options.packageDir)
      : join(options.repoRoot, "packages", "cli", "dist", "pkg");
  const outputPath =
    options.outputPath !== undefined
      ? resolveRepoPath(options.repoRoot, options.outputPath)
      : join(
          options.repoRoot,
          "packages",
          "cli",
          "dist",
          "pkg",
          "proxer-windows.msixbundle",
        );
  const bundleInputDir = join(
    options.repoRoot,
    "packages",
    "cli",
    "dist",
    "msix",
    "bundle",
  );
  const packageNames = (await readdir(packageDir))
    .filter((entry) => entry.endsWith(".msix"))
    .sort();

  if (packageNames.length === 0) {
    throw new Error(`No .msix packages found in ${packageDir}`);
  }

  await rm(bundleInputDir, { force: true, recursive: true });
  await mkdir(bundleInputDir, { recursive: true });

  await Promise.all(
    packageNames.map(async (packageName) => {
      await cp(
        join(packageDir, packageName),
        join(bundleInputDir, packageName),
      );
    }),
  );

  const makeAppxPath = await findWindowsSdkTool("makeappx");
  await execa(
    makeAppxPath,
    [
      "bundle",
      "/v",
      "/o",
      "/bv",
      version,
      "/d",
      bundleInputDir,
      "/p",
      outputPath,
    ],
    {
      cwd: options.repoRoot,
      stdio: "inherit",
    },
  );

  console.log(`MSIX bundle generated at ${outputPath}`);

  return { outputPath };
};

File: packages/tools/lib/src/lib/msix.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:archive/archive.dart';
import 'package:path/path.dart' as p;

import 'error.dart';
import 'exec.dart';
import 'version_files.dart';

const List<String> msixArchitectures = ['x64', 'arm64'];

String parseMsixArchitecture(String value) {
  if (!msixArchitectures.contains(value)) {
    throw ToolException(
      'Invalid MSIX architecture: $value. '
      'Must be one of: ${msixArchitectures.join(', ')}',
    );
  }

  return value;
}

class MsixIdentity {
  const MsixIdentity({
    required this.identityName,
    required this.publisher,
    required this.publisherDisplayName,
    this.displayName,
  });

  final String identityName;
  final String publisher;
  final String publisherDisplayName;
  final String? displayName;
}

const String _appDisplayName = 'proxer';
const String _appDescription =
    'Self-hosted reverse tunnel for HTTP, SSE, and WebSocket services';
const String _appExecutableName = 'proxer.exe';
const String _defaultLanguage = 'en-US';
const String _minWindowsVersion = '10.0.19041.0';
const String _maxTestedWindowsVersion = '10.0.26100.0';

String _resolveRepoPath(String repoRoot, String path) {
  return p.normalize(p.join(repoRoot, path));
}

String _readRequiredEnv(Map<String, String> env, String name) {
  final value = env[name];

  if (value == null || value.trim().isEmpty) {
    throw ToolException('$name is required to build Windows MSIX packages');
  }

  return value;
}

MsixIdentity readMsixIdentityFromEnv([Map<String, String>? environment]) {
  final env = environment ?? Platform.environment;
  final displayName = env['MSIX_DISPLAY_NAME'];

  return MsixIdentity(
    identityName: _readRequiredEnv(env, 'MSIX_IDENTITY_NAME'),
    publisher: _readRequiredEnv(env, 'MSIX_PUBLISHER'),
    publisherDisplayName: _readRequiredEnv(env, 'MSIX_PUBLISHER_DISPLAY_NAME'),
    displayName: displayName != null && displayName.trim().isNotEmpty
        ? displayName
        : null,
  );
}

final RegExp _digitsPattern = RegExp(r'^\d+$');

String convertVersionToMsixVersion(String version) {
  final coreVersion = version.split('-').first;
  final parts = coreVersion.split('.');

  if (parts.length != 3) {
    throw ToolException('Invalid package version for MSIX: $version');
  }

  final numericParts = <int>[];

  for (final part in parts) {
    if (!_digitsPattern.hasMatch(part)) {
      throw ToolException('Invalid package version for MSIX: $version');
    }

    final value = int.parse(part);

    if (value < 0 || value > 65535) {
      throw ToolException('MSIX version segment out of range: $part');
    }

    numericParts.add(value);
  }

  return '${numericParts[0]}.${numericParts[1]}.${numericParts[2]}.0';
}

Future<String> _readCliPackageVersion(String repoRoot) async {
  final pubspecPath = p.join(repoRoot, 'packages', 'cli', 'pubspec.yaml');
  final version = await readPubspecVersion(pubspecPath);

  return convertVersionToMsixVersion(version);
}

String _escapeXml(String value) {
  return value
      .replaceAll('&', '&amp;')
      .replaceAll('"', '&quot;')
      .replaceAll("'", '&apos;')
      .replaceAll('<', '&lt;')
      .replaceAll('>', '&gt;');
}

String buildMsixManifest({
  required String arch,
  required MsixIdentity identity,
  required String version,
}) {
  final displayName = identity.displayName ?? _appDisplayName;

  return '''
<?xml version="1.0" encoding="utf-8"?>
<Package
  xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10"
  xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10"
  xmlns:uap5="http://schemas.microsoft.com/appx/manifest/uap/windows10/5"
  xmlns:uap10="http://schemas.microsoft.com/appx/manifest/uap/windows10/10"
  xmlns:desktop4="http://schemas.microsoft.com/appx/manifest/desktop/windows10/4"
  xmlns:rescap="http://schemas.microsoft.com/appx/manifest/foundation/windows10/restrictedcapabilities"
  IgnorableNamespaces="uap uap5 uap10 desktop4 rescap">
  <Identity
    Name="${_escapeXml(identity.identityName)}"
    Publisher="${_escapeXml(identity.publisher)}"
    Version="${_escapeXml(version)}"
    ProcessorArchitecture="$arch" />
  <Properties>
    <DisplayName>${_escapeXml(displayName)}</DisplayName>
    <PublisherDisplayName>${_escapeXml(identity.publisherDisplayName)}</PublisherDisplayName>
    <Logo>Assets\\StoreLogo.png</Logo>
  </Properties>
  <Resources>
    <Resource Language="$_defaultLanguage" />
  </Resources>
  <Dependencies>
    <TargetDeviceFamily
      Name="Windows.Desktop"
      MinVersion="$_minWindowsVersion"
      MaxVersionTested="$_maxTestedWindowsVersion" />
  </Dependencies>
  <Capabilities>
    <rescap:Capability Name="runFullTrust" />
  </Capabilities>
  <Applications>
    <Application
      Id="proxer"
      Executable="$_appExecutableName"
      EntryPoint="Windows.FullTrustApplication"
      uap10:RuntimeBehavior="packagedClassicApp"
      uap10:TrustLevel="mediumIL"
      desktop4:SupportsMultipleInstances="true">
      <uap:VisualElements
        DisplayName="${_escapeXml(displayName)}"
        Description="${_escapeXml(_appDescription)}"
        Square150x150Logo="Assets\\Square150x150Logo.png"
        Square44x44Logo="Assets\\Square44x44Logo.png"
        BackgroundColor="#102A43" />
      <Extensions>
        <uap5:Extension Category="windows.appExecutionAlias">
          <uap5:AppExecutionAlias desktop4:Subsystem="console">
            <uap5:ExecutionAlias Alias="$_appExecutableName" />
          </uap5:AppExecutionAlias>
        </uap5:Extension>
      </Extensions>
    </Application>
  </Applications>
</Package>
''';
}

final List<int> _crcTable = List<int>.generate(256, (index) {
  var crc = index;

  for (var bit = 0; bit < 8; bit++) {
    crc = (crc & 1) == 1 ? 0xedb88320 ^ (crc >> 1) : crc >> 1;
  }

  return crc;
});

int _crc32(List<List<int>> buffers) {
  var crc = 0xffffffff;

  for (final buffer in buffers) {
    for (final byte in buffer) {
      crc = (crc >> 8) ^ _crcTable[(crc ^ byte) & 0xff];
    }
  }

  return (crc ^ 0xffffffff) & 0xffffffff;
}

Uint8List _pngChunk(String type, List<int> data) {
  final typeBytes = ascii.encode(type);
  final chunk = BytesBuilder(copy: false);
  final length = ByteData(4)..setUint32(0, data.length);
  final crc = ByteData(4)..setUint32(0, _crc32([typeBytes, data]));

  chunk.add(length.buffer.asUint8List());
  chunk.add(typeBytes);
  chunk.add(data);
  chunk.add(crc.buffer.asUint8List());

  return chunk.takeBytes();
}

Uint8List _createLogoPng(int size) {
  const bytesPerPixel = 4;
  final rowSize = 1 + size * bytesPerPixel;
  final raw = Uint8List(rowSize * size);
  const base = [16, 42, 67, 255];
  const accent = [20, 184, 166, 255];
  const light = [245, 247, 250, 255];
  final bandStart = (size * 0.22).floor();
  final bandEnd = (size * 0.34).floor();
  final markStart = (size * 0.42).floor();
  final markEnd = (size * 0.58).floor();

  for (var y = 0; y < size; y++) {
    final rowOffset = y * rowSize;
    raw[rowOffset] = 0;

    for (var x = 0; x < size; x++) {
      final pixelOffset = rowOffset + 1 + x * bytesPerPixel;
      final List<int> color;

      if ((x >= bandStart && x <= bandEnd) ||
          (x >= markStart && x <= markEnd && y >= markStart && y <= markEnd)) {
        color = accent;
      } else if (x >= markStart && x <= markEnd) {
        color = light;
      } else {
        color = base;
      }

      raw[pixelOffset] = color[0];
      raw[pixelOffset + 1] = color[1];
      raw[pixelOffset + 2] = color[2];
      raw[pixelOffset + 3] = color[3];
    }
  }

  final signature = Uint8List.fromList([
    0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a, //
  ]);
  final ihdr = ByteData(13)
    ..setUint32(0, size)
    ..setUint32(4, size);
  final ihdrBytes = ihdr.buffer.asUint8List();
  ihdrBytes[8] = 8;
  ihdrBytes[9] = 6;
  ihdrBytes[10] = 0;
  ihdrBytes[11] = 0;
  ihdrBytes[12] = 0;

  final png = BytesBuilder(copy: false);
  png.add(signature);
  png.add(_pngChunk('IHDR', ihdrBytes));
  png.add(_pngChunk('IDAT', ZLibEncoder().encode(raw)));
  png.add(_pngChunk('IEND', const []));

  return png.takeBytes();
}

Future<void> _writeMsixAssets(String assetsDirectory) async {
  await Directory(assetsDirectory).create(recursive: true);

  await File(
    p.join(assetsDirectory, 'StoreLogo.png'),
  ).writeAsBytes(_createLogoPng(50));
  await File(
    p.join(assetsDirectory, 'Square44x44Logo.png'),
  ).writeAsBytes(_createLogoPng(44));
  await File(
    p.join(assetsDirectory, 'Square150x150Logo.png'),
  ).writeAsBytes(_createLogoPng(150));
}

final RegExp _naturalChunkPattern = RegExp(r'(\d+)|(\D+)');

int _compareNatural(String left, String right) {
  final leftChunks = _naturalChunkPattern.allMatches(left).toList();
  final rightChunks = _naturalChunkPattern.allMatches(right).toList();
  final length = leftChunks.length < rightChunks.length
      ? leftChunks.length
      : rightChunks.length;

  for (var index = 0; index < length; index++) {
    final leftDigits = leftChunks[index].group(1);
    final rightDigits = rightChunks[index].group(1);
    int comparison;

    if (leftDigits != null && rightDigits != null) {
      comparison = int.parse(leftDigits).compareTo(int.parse(rightDigits));
    } else {
      comparison = leftChunks[index]
          .group(0)!
          .compareTo(rightChunks[index].group(0)!);
    }

    if (comparison != 0) {
      return comparison;
    }
  }

  return leftChunks.length.compareTo(rightChunks.length);
}

Future<String> _findWindowsSdkTool(String toolName) async {
  final envOverride = Platform.environment['${toolName.toUpperCase()}_PATH'];

  if (envOverride != null && envOverride.trim().isNotEmpty) {
    return envOverride;
  }

  if (!Platform.isWindows) {
    throw ToolException('$toolName.exe is only available on Windows runners');
  }

  final windowsSdkDir = Platform.environment['WindowsSdkDir'];
  final programFiles = Platform.environment['ProgramFiles'];
  final programFilesX86 = Platform.environment['ProgramFiles(x86)'];

  final sdkRoots = <String>[
    if (windowsSdkDir != null) p.join(windowsSdkDir, 'bin'),
    if (programFilesX86 != null)
      p.join(programFilesX86, 'Windows Kits', '10', 'bin'),
    if (programFiles != null) p.join(programFiles, 'Windows Kits', '10', 'bin'),
  ];

  for (final sdkRoot in sdkRoots) {
    List<String> versions;

    try {
      versions =
          Directory(sdkRoot)
              .listSync()
              .whereType<Directory>()
              .map((entry) => p.basename(entry.path))
              .toList()
            ..sort((left, right) => _compareNatural(right, left));
    } on FileSystemException {
      continue;
    }

    for (final version in versions) {
      for (final arch in ['x64', 'x86']) {
        final candidate = p.join(sdkRoot, version, arch, '$toolName.exe');

        if (File(candidate).existsSync()) {
          return candidate;
        }
      }
    }
  }

  return '$toolName.exe';
}

Future<void> _createPriFile(String packageRoot) async {
  final makePriPath = await _findWindowsSdkTool('makepri');
  final priConfigPath = p.join(packageRoot, 'priconfig.xml');

  await runInherited(makePriPath, [
    'createconfig',
    '/cf',
    priConfigPath,
    '/dq',
    _defaultLanguage,
  ], workingDirectory: packageRoot);

  await runInherited(makePriPath, [
    'new',
    '/pr',
    packageRoot,
    '/cf',
    priConfigPath,
  ], workingDirectory: packageRoot);

  final priConfigFile = File(priConfigPath);

  if (priConfigFile.existsSync()) {
    await priConfigFile.delete();
  }
}

Future<void> _removeDirectory(String path) async {
  final directory = Directory(path);

  if (directory.existsSync()) {
    await directory.delete(recursive: true);
  }
}

Future<({String outputPath, String packageRoot})> performMsixBuild({
  required String arch,
  required String executablePath,
  required String repoRoot,
  String? outputPath,
  String? packageRoot,
}) async {
  final identity = readMsixIdentityFromEnv();
  final version = await _readCliPackageVersion(repoRoot);
  final resolvedPackageRoot = packageRoot != null
      ? _resolveRepoPath(repoRoot, packageRoot)
      : p.join(repoRoot, 'packages', 'cli', 'dist', 'msix', arch);
  final resolvedOutputPath = outputPath != null
      ? _resolveRepoPath(repoRoot, outputPath)
      : p.join(
          repoRoot,
          'packages',
          'cli',
          'dist',
          'pkg',
          'proxer-win-$arch.msix',
        );
  final sourceExecutablePath = _resolveRepoPath(repoRoot, executablePath);

  await _removeDirectory(resolvedPackageRoot);
  await Directory(resolvedPackageRoot).create(recursive: true);
  await Directory(p.dirname(resolvedOutputPath)).create(recursive: true);
  await File(
    sourceExecutablePath,
  ).copy(p.join(resolvedPackageRoot, _appExecutableName));
  await File(p.join(resolvedPackageRoot, 'AppxManifest.xml')).writeAsString(
    buildMsixManifest(arch: arch, identity: identity, version: version),
  );
  await _writeMsixAssets(p.join(resolvedPackageRoot, 'Assets'));
  await _createPriFile(resolvedPackageRoot);

  final makeAppxPath = await _findWindowsSdkTool('makeappx');

  await runInherited(makeAppxPath, [
    'pack',
    '/v',
    '/o',
    '/h',
    'SHA256',
    '/d',
    resolvedPackageRoot,
    '/p',
    resolvedOutputPath,
  ], workingDirectory: repoRoot);

  stdout.writeln('MSIX package generated at $resolvedOutputPath');

  return (outputPath: resolvedOutputPath, packageRoot: resolvedPackageRoot);
}

Future<({String outputPath})> performMsixBundle({
  required String repoRoot,
  String? outputPath,
  String? packageDir,
}) async {
  final version = await _readCliPackageVersion(repoRoot);
  final resolvedPackageDir = packageDir != null
      ? _resolveRepoPath(repoRoot, packageDir)
      : p.join(repoRoot, 'packages', 'cli', 'dist', 'pkg');
  final resolvedOutputPath = outputPath != null
      ? _resolveRepoPath(repoRoot, outputPath)
      : p.join(
          repoRoot,
          'packages',
          'cli',
          'dist',
          'pkg',
          'proxer-windows.msixbundle',
        );
  final bundleInputDir = p.join(
    repoRoot,
    'packages',
    'cli',
    'dist',
    'msix',
    'bundle',
  );
  final packageNames =
      Directory(resolvedPackageDir)
          .listSync()
          .whereType<File>()
          .map((entry) => p.basename(entry.path))
          .where((name) => name.endsWith('.msix'))
          .toList()
        ..sort();

  if (packageNames.isEmpty) {
    throw ToolException('No .msix packages found in $resolvedPackageDir');
  }

  await _removeDirectory(bundleInputDir);
  await Directory(bundleInputDir).create(recursive: true);

  for (final packageName in packageNames) {
    await File(
      p.join(resolvedPackageDir, packageName),
    ).copy(p.join(bundleInputDir, packageName));
  }

  final makeAppxPath = await _findWindowsSdkTool('makeappx');

  await runInherited(makeAppxPath, [
    'bundle',
    '/v',
    '/o',
    '/bv',
    version,
    '/d',
    bundleInputDir,
    '/p',
    resolvedOutputPath,
  ], workingDirectory: repoRoot);

  stdout.writeln('MSIX bundle generated at $resolvedOutputPath');

  return (outputPath: resolvedOutputPath);
}
